### PR TITLE
Cuda mr cache

### PR DESCRIPTION
**Pull request by iziemba**
PR to enable CUDA MR cache. Consists of two components:
~~1. Update MR cache to support multiple monitors.~~
2. Define CUDA memory monitor.

Needs ~~https://github.com/ofiwg/libfabric/pull/5976~~, https://github.com/ofiwg/libfabric/pull/6001

### CODE REVIEW

**Comment by iziemba** [comment]
If it would be easier to future break up this PR, let me know.

**Comment by rajachan** [comment]
EFA failures are real, segfaults in the regattr unit tests:

```
server_cmd: /home/ubuntu/libfabric/fabtests/install/bin/fi_mr_test -p "efa"
server_stdout: |
Testing MR on fabric EFA-fe80::aa:55ff:fe0b:81a4
Running mr_reg [Test fi_mr_reg across different access combinations]...PASS!
Running mr_regv [Test fi_mr_regv across various buffer sizes]...PASS!
Running mr_regattr [Test fi_mr_regattr across various buffer sizes]...bash: line 1: 62464 Segmentation fault (core dumped) /home/ubuntu/libfabric/fabtests/install/bin/fi_mr_test -p "efa"
```

Will check out what's happening.

**Comment by rajachan** [comment]
Was able to reproduce it:

```
#0  0x00007f079446bf4a in ofi_mr_cache_search (cache=0x947118, attr=0x7ffda60c76a0, entry=0x7ffda60c7590) at prov/util/src/util_mr_cache.c:299
299		struct ofi_mem_monitor *monitor = cache->monitors[attr->iface];
[Current thread is 1 (Thread 0x7f07949bd740 (LWP 83714))]
Missing separate debuginfos, use: debuginfo-install glibc-2.26-34.amzn2.x86_64 libatomic-7.3.1-6.amzn2.0.4.x86_64 libibverbs-27.0-1.amzn2.x86_64 libnl3-3.2.28-4.amzn2.0.1.x86_64 librdmacm-27.0-1.amzn2.x86_64
(gdb) bt
#0  0x00007f079446bf4a in ofi_mr_cache_search (cache=0x947118, attr=0x7ffda60c76a0, entry=0x7ffda60c7590) at prov/util/src/util_mr_cache.c:299
#1  0x00007f07944cccff in efa_mr_cache_regattr (fid=0x947030, attr=0x7ffda60c76a0, flags=0, mr_fid=0x7ffda60c7688) at prov/efa/src/efa_mr.c:169
#2  0x00007f07944d017c in fi_mr_regattr (domain=0x947030, attr=0x7ffda60c76a0, flags=0, mr=0x7ffda60c7688) at ./include/rdma/fi_domain.h:346
#3  0x00007f07944d02ad in rxr_mr_regattr (domain_fid=0x946c70, attr=0x7ffda60c76a0, flags=0, mr=0x7ffda60c7688) at prov/efa/src/rxr/rxr_domain.c:103
#4  0x00000000004017e2 in fi_mr_regattr (domain=0x946c70, attr=0x7ffda60c76a0, flags=0, mr=0x7ffda60c7688) at /shared/libfabric/install//include/rdma/fi_domain.h:346
#5  0x0000000000401df1 in mr_regattr () at unit/mr_test.c:172
#6  0x00000000004023a9 in run_tests (test_array=0x613200 <test_array>, err_buf=0x613680 <err_buf> "") at unit/common.c:66
#7  0x00000000004021d3 in main (argc=3, argv=0x7ffda60c7848) at unit/mr_test.c:263
(gdb) p *cache
$1 = {domain = 0x947030, monitors = {0x7f07947b2120 <uffd>, 0x0}, notify_entries = {{next = 0x7f07947b2120 <uffd>, prev = 0x7f07947b2120 <uffd>}, {next = 0x0, prev = 0x0}}, entry_data_size = 88, storage = {type = OFI_MR_STORAGE_DEFAULT,
    storage = 0x947b50, find = 0x7f079446c887 <ofi_mr_rbt_find>, overlap = 0x7f079446c8ca <ofi_mr_rbt_overlap>, insert = 0x7f079446c914 <ofi_mr_rbt_insert>, erase = 0x7f079446c976 <ofi_mr_rbt_erase>,
    destroy = 0x7f079446c868 <ofi_mr_rbt_destroy>}, lru_list = {next = 0x94a198, prev = 0x94a198}, flush_list = {next = 0x947fd8, prev = 0x94a0d8}, lock = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = 0,
      __spins = 0, __elision = 0, __list = {__prev = 0x0, __next = 0x0}}, __size = '\000' <repeats 39 times>, __align = 0}, cached_cnt = 1, cached_size = 8388608, uncached_cnt = 0, uncached_size = 0, search_cnt = 185, delete_cnt = 185,
  hit_cnt = 139, notify_cnt = 0, entry_pool = 0x947cd0, add_region = 0x7f07944cc9de <efa_mr_cache_entry_reg>, delete_region = 0x7f07944ccac7 <efa_mr_cache_entry_dereg>}
(gdb) p *cache->monitors
$2 = (struct ofi_mem_monitor *) 0x7f07947b2120 <uffd>
(gdb) p *attr
$3 = {mr_iov = 0x945190, iov_count = 1, access = 12288, offset = 139670510311661, requested_key = 49374, context = 0x0, auth_key_size = 4253443, auth_key = 0x7f07935a3760 <_IO_2_1_stdout_> "\204*\255", <incomplete sequence \373>,
  iface = (unknown: 2472145920), device = {reserved = 139670510268637, cuda = -1826197283}}
```

That attr->iface doesn't look right.

**Comment by rajachan** [comment]
> That attr->iface doesn't look right.

https://github.com/ofiwg/libfabric/pull/5984 fixes it.

File: prov/util/src/cuda_mem_monitor.c [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+#ifdef HAVE_LIBCUDA
+
+#include "ofi_hmem.h"
+
+static int cuda_mm_subscribe(struct ofi_mem_monitor *notifier, const void *addr,
+			     size_t len, union ofi_mr_hmem_info *hmem_info)
+{
+	CUresult ret;
+
+	ret = ofi_cuPointerGetAttribute(&hmem_info->cuda_id,
```
Comment (rajachan) [inline prov/util/src/cuda_mem_monitor.c:44]: I had this question in https://github.com/ofiwg/libfabric/issues/5789#issuecomment-608640709 but we did not get to talking about it. If all we need is the buf ID, should we consider extending the device union in fi_mr_attr to include this information? (Thinking about the ~3µs overhead this is going to introduce on each subscribe call).

Comment (iziemba) [reply prov/util/src/cuda_mem_monitor.c:44]: One thing which is not 100% clear is when the buffer ID needs to be allocated. The Nvidia RDMA GPU direct documentation seems to imply that the buffer ID needs to be allocated after device pages are pinned. 

https://docs.nvidia.com/cuda/gpudirect-rdma/index.html#invalidating-based-on-buffer-id

**Comment by iziemba** [comment]
I am going to break this PR up: this will remain the CUDA MR cache. The new PR will only be for multiple memory monitors per cache.

**Comment by jswaro** [comment]
@shefty @iziemba What is the state of this PR?

**Comment by shefty** [comment]
I believe this needs PR #6001 merged first.

**Comment by iziemba** [comment]
> I believe this needs PR #6001 merged first.

Correct. PR has been rebased and is ready for review.

File: include/ofi_mr.h [hunk 2]
```diff
@@ -177,6 +177,8 @@ struct ofi_memhooks {
 
 extern struct ofi_mem_monitor *memhooks_monitor;
 
+/* Cuda memory monitor. */
```
Comment (shefty) [inline include/ofi_mr.h:181]: nit: comment is obvious

File: include/ofi_mr.h [hunk 3]
```diff
@@ -241,6 +242,7 @@ struct ofi_mr_cache_params {
 	size_t				max_cnt;
 	size_t				max_size;
 	char *				monitor;
+	int				cuda_monitor_enabled;
```
Comment (shefty) [inline include/ofi_mr.h:245]: There should be a more general way of marking each monitor as enabled/disabled.  Something that applies to memhooks, uffd, etc.

**Comment by shefty** [comment]
bot:aws:retest

**Comment by shefty** [comment]
Once we verify with CI, I'm okay merging these changes as-is.  We can figure out to how generalize the monitor enabling mechanism separately.

File: include/ofi_mr.h (hunk 2 above)
Comment (iziemba) [reply include/ofi_mr.h:181]: I'll remove.

**Comment by iziemba** [comment]
Any other changes needed for this PR?

**Comment by iziemba** [comment]
Thanks!

**Comment by PeterGottesman** [comment]
@iziemba Do you have/know of some test that can be used to validate CUDA MR cache correctness?

**Comment by iziemba** [comment]
@PeterGottesman I have an updated version of fi_mr_cache_evict which I have been using to verify CUDA MR cache functionality. But, this test is reliant on a provider supporting the MR cache with CUDA memory. Currently, no public providers do. I have been holding on opening a PR for these changes until CUDA support has landed for the verbs provider.

**Comment by shefty** [comment]
@iziemba - FWIW, I'm okay reviewing and merging changes to support an out of tree provider, as long as the changes themselves are generic enough to apply to any.

**Comment by iziemba** [comment]
Sounds good. I'll open the PR.

**Comment by PeterGottesman** [comment]
Awesome, thanks @iziemba

**Pull request merged**